model/tests/test_lif.py

```diff
@@ -0,0 +1,61 @@
+import torch
+
+from model.src.lif import LIF
+
+
+def test_lif_spikes_for_membrane_above_0() -> None:
+    lif = LIF(beta=0.9, threshold=1.0)
+
+    # make sure no spike at first
+    input = torch.tensor([[2.0, 2.0, 2.0], [2.0, 2.0, 2.0]])
+    spk = lif(input)
+    assert spk.shape == input.shape
+    assert (spk == 0).all()  # type: ignore
```

Comment:
@mstormbull This behavior (spk == 0 even when current is way above threshold)  is present due to the update ordering for the spike in your sample code. You had the spike firing at the next timestep after the mem > threshold. Is that what we want though?

This is an implementation detail but it leaks into the dynamics of the learning rule and becomes very important. I feel that the learning rule is meant to be local in space and time, but that doesn't necessarily determine we should do it one way vs another.

Reply:
You are 100% correct. I've always done the next timestep for the reason that we want to current spike state to influence our current update equation. I suppose we could discuss switching this, but my impression is that you would like to calculate the spikes based upon the previous timestep -- my view is this is analogous to taking the nonlinearity applied to your synaptic drive in the rate model. Do you see any unanticipated consequences of that choice?

Reply:
We could of course report back the spikes during the current timestep where you exceed threshold and then call that stored value for the update equation. I feel like that accounts for your spk==0 concern and doesn't change the dynamics, but it does necessitate the need to expand the scope of the spk variable by saving it as self.spk. Correct me if I am wrong about that.

Reply:
> I suppose we could discuss switching this, but my impression is that you would like to calculate the spikes based upon the previous timestep -- my view is this is analogous to taking the nonlinearity applied to your synaptic drive in the rate model. 

In the rate model I believe the per timestep logic works differently than what we are implementing here. Every layer stores an activation, then in a single timestep:
1. The synaptic drive is calculated for all connections into that layer.
2. The non linearity is applied
3. The layer gets a new activation state
4. Weight update rule is applied based on the new activations (but really the weight update rule is more complicated and contains gating factor, presynaptic current, and post synaptic firing rate -- image below). 


![image](https://github.com/and-rewsmith/LPL-SNN/assets/19913741/833536a2-0156-4511-a51f-269a6f6f22ac)


In this current implementation for spiking network it is staggered relative to above:
1. The neuron assesses whether it will spike or not and issues spike (akin to step 3 above)
2. The membrane potential decays (and reset potentially applied)
3. The learning rule updates the synaptic weights based on the decayed membrane potential and the spike before decay. 

If we want to be sure the spiking rule maps to the rate coded network we should be sure that having the spike be issued at the previous timestep maps well to the weight update rule for the rate coded network, notably containing post synaptic "spike" (firing rate) for both current timestep and previous timestep, which I think makes thing a bit ambiguous.

Alternatively, it may be fine to diverge from what the rate coded model did and follow more closely what Zenke did. He has a github repo for the actual LPL rule in the paper which we can look at to see exactly how he is implementing the update rule if it comes to that.

There is also an erratum I was taking a look at. Will need to issue a PR to address some of that.

Reply:
Yes, good comment. I can't foresee where this decision is going to come back to bite us and I don't have a normatively founded opinion on it. With that, my suggestion is to dive the Zenke implementation, but let's discuss.

Reply:
Regarding the erratum, issue 5: we anticipated from our calculations and reference to 2011 science. They did not issue the 'j->k'-fix, but I suspect we are correct about that (also supported by the 2011 SI).

Reply:
The fix to the double filter is also just a scaling. Important but not conceptually troublesome.